### ISSUE TITLE
fix(neon): make weight resolution validate raw tensor extent before repacking

### ISSUE THREAD

**Issue by geisten** [geisten/geistlib#325]
Severity: HIGH

## What happened

The ASan unit `tests/test_resolve_weight_unit.c:94-107` supplies a 64-byte raw buffer for every dtype. The Q4_K resolver begins a full repack in `src/backends/cpu_neon/weight_resolve.c:1227-1258` and reads beyond that buffer.

Disabling Q4_K predecode avoids the failure:

```sh
GEIST_Q4K_PREDECODE=0 bin/mac-omp/asan/tests/test_resolve_weight_unit
```

The deeper API problem is that `struct geist_weight` (`include/geist_weight.h:134-152`) has dimensions and a raw pointer but no `raw_nbytes`, so the resolver cannot independently verify the source extent before conversion.

## Expected

Carry the raw tensor extent into resolution and reject any layout conversion whose required bytes exceed it. Fix the unit fixture to allocate format-correct storage; do not merely enlarge it without adding production validation.

## Acceptance criteria

- `struct geist_weight` or an equivalent resolver input exposes the raw extent.
- Every repack/predecode path validates source and destination sizes with checked arithmetic.
- The resolver ASan test passes with Q4_K predecode enabled.
- Add deliberately short buffers for each repacked dtype and assert a clean error/fallback.

## Mandatory C23 implementation rules

- In every modified array API, place all length/capacity parameters before the arrays they describe; use VLA contracts such as `T data[static len]` where the non-null/minimum-size precondition is real.
- Use C23 `bool`, `true`, `false`, and `nullptr`/`nullptr_t` instead of integer booleans or `NULL`. Replace constant macros with typed `constexpr` objects when compile-time and ABI requirements permit.
- Use `[[nodiscard]]` for returned status/size values that must be checked. Add `restrict` only when non-aliasing is proven for every caller. Use `typeof`/`auto` only when they improve clarity without hiding ownership or changing code generation.
- Perform checked `size_t` arithmetic before allocation, indexing, alignment, or pointer advancement. For cursors, validate against `(size_t)(end - p)` before advancing; do not use `p + len` or `&p[len]` as the bounds check.
- Preserve public ABI unless the ticket explicitly includes a versioned compatibility migration.

## Mandatory performance and quality gates

- Capture a baseline before the first change and rerun after every resolver/repack change that can affect loading or inference.
- Measure model load time/RSS, `bench_q4k_kernel`, and the standard macOS prefill/decode sweep with identical compiler, model hash, thread settings, and environment; retain raw results.
- Resolver parity tests and full-model logits must remain identical within the existing tolerance; do not widen tolerances.
- Throughput and latency must not be worse than baseline beyond the predeclared measurement-noise band. Any repeatable regression blocks acceptance.

**Comment by geisten** [comment]
Fixed in c830bbd (PR #344), released in **v0.10.2**.

**What changed.** `struct geist_weight` gains `raw_nbytes` — the bytes actually readable at `raw`, deliberately **not** derived from `(dtype, n_in, n_out)`, since the two disagree exactly when something is wrong. `quant.h` gains `quant_raw_bytes()` (the required extent for a dtype and element count, checked arithmetic, next to the block constants so a layout change and a size change are one edit) and `quant_weight_extent_ok()` on top of it.

All four resolvers — `cpu_neon`, `cpu_scalar`, `metal`, `vulkan` — reject a zero `raw_nbytes` with `GEIST_E_INVALID_ARG` and a short one with `GEIST_E_FORMAT`, before installing a kernel or repacking. **Vulkan was the same bug in different clothes**: `vk_weight_bytes(w)` derives the upload size from the shape and reads that many bytes out of `raw`. `geist_weight` is `EXPERIMENTAL`, so the added field is in scope.

**Acceptance criteria.**
- [x] Resolver input exposes the raw extent
- [x] Every repack/predecode path validates source and destination with checked arithmetic
- [x] The resolver ASan test passes **with Q4_K predecode enabled**
- [x] Short buffers per repacked dtype assert a clean error — one byte under the required extent, for each of the twelve dtypes with a fixed raw layout, refused with no kernel installed and no aux allocated
- [x] Fixture allocates format-correct storage (was 64 bytes for every dtype with a Q4_K special case) rather than merely being enlarged

**Drift guard.** `test_resolve_weight_unit` now sweeps the whole dtype enum and fails if `cpu_neon` can install a kernel for a dtype `quant_raw_bytes` cannot size. That gap opened silently once already — IQ4_NL/IQ4_XS landed on main after this contract was written, closed in 8cb8d5a.

**Performance gate.** Load-time only. Model load 1.48 s best-of-7; Q4_K/Q6_K microkernels unchanged within run-to-run spread; sweep in PR #344 shows no regression.
